iceoryx_hoofs/include/iceoryx_hoofs/internal/cxx/string.inl

```diff
@@ -214,61 +214,222 @@ inline bool string<Capacity>::unsafe_assign(const std::string& str) noexcept
+                            bool>::type
+    string<Capacity>::operator<(const T& rhs) const noexcept
+{
+    return (compare(rhs) < 0);
+}
+
+template <uint64_t Capacity>
+template <typename T>
+inline
+    typename std::enable_if<(is_cxx_string<T>::value || is_char_array<T>::value || std::is_same<T, std::string>::value),
+                            bool>::type
+    string<Capacity>::operator<=(const T& rhs) const noexcept
+{
+    return (compare(rhs) <= 0);
+}
+
+template <uint64_t Capacity>
+template <typename T>
+inline
+    typename std::enable_if<(is_cxx_string<T>::value || is_char_array<T>::value || std::is_same<T, std::string>::value),
+                            bool>::type
+    string<Capacity>::operator>(const T& rhs) const noexcept
+{
+    return (compare(rhs) > 0);
+}
+
+template <uint64_t Capacity>
+template <typename T>
+inline
+    typename std::enable_if<(is_cxx_string<T>::value || is_char_array<T>::value || std::is_same<T, std::string>::value),
+                            bool>::type
+    string<Capacity>::operator>=(const T& rhs) const noexcept
+{
+    return (compare(rhs) >= 0);
+}
+
+template <typename T, uint64_t Capacity>
+inline typename std::enable_if<(is_char_array<T>::value || std::is_same<T, std::string>::value), bool>::type
+operator==(const T& lhs, const string<Capacity>& rhs) noexcept
+{
+    return (rhs.compare(lhs) == 0);
+}
+
+template <typename T, uint64_t Capacity>
+inline typename std::enable_if<(is_char_array<T>::value || std::is_same<T, std::string>::value), bool>::type
+operator!=(const T& lhs, const string<Capacity>& rhs) noexcept
+{
+    return (rhs.compare(lhs) != 0);
+}
+
+template <typename T, uint64_t Capacity>
+inline typename std::enable_if<(is_char_array<T>::value || std::is_same<T, std::string>::value), bool>::type
+operator<(const T& lhs, const string<Capacity>& rhs) noexcept
+{
+    return (rhs.compare(lhs) > 0);
+}
+
+template <typename T, uint64_t Capacity>
+inline typename std::enable_if<(is_char_array<T>::value || std::is_same<T, std::string>::value), bool>::type
+operator<=(const T& lhs, const string<Capacity>& rhs) noexcept
+{
+    return (rhs.compare(lhs) >= 0);
+}
+
+template <typename T, uint64_t Capacity>
+inline typename std::enable_if<(is_char_array<T>::value || std::is_same<T, std::string>::value), bool>::type
+operator>(const T& lhs, const string<Capacity>& rhs) noexcept
+{
+    return (rhs.compare(lhs) < 0);
+}
+
+template <typename T, uint64_t Capacity>
+inline typename std::enable_if<(is_char_array<T>::value || std::is_same<T, std::string>::value), bool>::type
+operator>=(const T& lhs, const string<Capacity>& rhs) noexcept
+{
+    return (rhs.compare(lhs) <= 0);
+}
+
+template <uint64_t Capacity>
+inline int64_t string<Capacity>::compare(const char& other) const noexcept
+{
+    auto result = memcmp(c_str(), &other, 1U);
+    if (result == 0)
     {
-        return 1;
+        if (m_rawstringSize > 1U)
+        {
+            return 1;
+        }
+        else
```

Comment:
The `< 1U` case would be the empty string case and for this we would always correctly compare -1 if `other != \0`. Hence it appears we can omit the case, but there is the following problem:

If `this` is the empty string (size 0) and `other` is  `\0` we would get result 0 and also return 0 which is wrong. I may miss something here but I am fairly certain.

Lexicographically the empty string is smaller than any character, in particular `\0` (which is not equal to the empty string). I think the logic is incorrect in this case and we need to properly handle the empty string. The easiest would be an initial size check (there may be more efficient ways in terms of control flow).

Reply:
Actually the first character of an empty string is `\0`. It is default constructed this way and it is also set in the clear method. 
For me, it makes sense to return 0 in this case. Furthermore, if you compare an empty std::string with `"\0"` (note that a char comparison is not allowed) it also returns 0.

Reply:
That may be the case for `std::string` and we can leave it like this. But in the theory of formal languages (https://en.wikipedia.org/wiki/Empty_string#:~:text=In%20formal%20treatments%2C%20the%20empty,%7C%CE%B5%7C%20%3D%200.) this makes no sense and teh STL is at odds with this.

We can leave it as is for STL consistency.

Reply:
@MatthiasKillat @elfenpiff I reconsidered this case and decided to differ here from the std::string behavior. If an empty cxx::string is compared with '\0', compare will return -1.